hazelcast/sql.py

```diff
@@ -0,0 +1,1604 @@
+
+                self._execute_response.set_result(response)
+            else:
+                # Result only contains the update count.
+                response = _ExecuteResponse(None, None, update_count)
+                self._execute_response.set_result(response)
+
+                # There is nothing more we can get from the server.
+                self._mark_closed()
+
+    def _mark_closed(self):
+        """Marks the result as closed."""
+        with self._lock:
+            self._closed = True
+
+    def __enter__(self):
+        # The execute request is already sent.
+        # There is nothing more to do.
+        return self
+
+    def __exit__(self, *_):
+        # Ignoring the possible exception details
+        # since we close the query regardless of that.
+        self.close().result()
+
+
+class _InternalSqlService(object):
+    """Internal SQL service that offers more public API
+    than the one exposed to the user.
+    """
+
+    def __init__(self, connection_manager, serialization_service, invocation_service):
+        self._connection_manager = connection_manager
+        self._serialization_service = serialization_service
+        self._invocation_service = invocation_service
+
+    def execute(self, sql, *params):
+        """Constructs a statement and executes it.
+
+        Args:
+            sql (str): SQL string.
+            *params: Query parameters.
+
+        Returns:
+            SqlResult: The execution result.
+        """
+        statement = SqlStatement(sql)
+
+        for param in params:
+            statement.add_parameter(param)
+
+        return self.execute_statement(statement)
+
+    def execute_statement(self, statement):
+        """Executes the given statement.
+
+        Args:
+            statement (SqlStatement): The statement to execute.
+
+        Returns:
+            SqlResult: The execution result.
+        """
+
+        # Get a random Data member (non-lite member)
+        connection = self._connection_manager.get_random_connection(True)
+        if not connection:
+            # Either the client is not connected to the cluster, or
+            # there are no data members in the cluster.
+            raise HazelcastSqlError(
+                self.get_client_id(),
+                _SqlErrorCode.CONNECTION_PROBLEM,
+                "Client is not currently connected to the cluster.",
+                None,
+            )
+
+        # Create a new, unique query id.
+        query_id = _SqlQueryId.from_uuid(connection.remote_uuid)
+
+        # Serialize the passed parameters.
+        serialized_params = []
+        for param in statement.parameters:
+            serialized_params.append(self._serialization_service.to_data(param))
+
+        request = sql_execute_codec.encode_request(
+            statement.sql,
+            serialized_params,
+            # to_millis expects None to produce -1
+            to_millis(None if statement.timeout == -1 else statement.timeout),
```

Comment:
How about `-1 if statement.timeout == -1 else statement.timeout*1000` instead of `to_millis( ... )`?

Reply:
I think the current version is better because I know (from the other places), we use None for unset timeout variables. But if you think we really need to change this I can do it